domain/modelagent/state/modelstate.go

```diff
@@ -137,35 +194,126 @@ func (st *State) GetMachineUUID(ctx context.Context, name machine.Name) (string,
+		err := st.checkMachineNotDead(ctx, tx, uuid)
+		if errors.Is(err, machineerrors.MachineNotFound) {
+			return errors.Errorf(
+				"machine %q does not exist", uuid,
+			).Add(machineerrors.MachineNotFound)
+		} else if err != nil && !errors.Is(err, machineerrors.MachineIsDead) {
+			return errors.Errorf(
+				"checking machine %q exists: %w", uuid, err,
+			)
+		}
+
+		err = tx.Query(ctx, stmt, machineUUID).Get(&rval)
+		if errors.Is(err, sql.ErrNoRows) {
+			return errors.Errorf(
+				"machine %q has no agent version set", uuid,
+			).Add(modelagenterrors.AgentVersionNotFound)
+		} else if err != nil {
+			return errors.Errorf(
+				"getting machine %q agent version: %w",
+				uuid, err,
+			)
+		}
+		return nil
+	})
+	if err != nil {
+		return coreagentbinary.Version{}, errors.Capture(err)
+	}
+
+	vers, err := semversion.Parse(rval.Version)
+	if err != nil {
+		return coreagentbinary.Version{}, errors.Errorf(
+			"parsing machine %q agent version: %w",
+			uuid, err,
+		)
+	}
+	return coreagentbinary.Version{
+		Number: vers,
+		Arch:   rval.Architecture,
+	}, nil
+}
+
+// GetMachineTargetAgentVersion returns the target agent version for the
+// specified machine.
 // The following error types can be expected:
-// - [modelagenterrors.AgentVersionNotFound] - when the agent version does not exist.
+// - [modelagenterrors.AgentVersionNotFound] when the agent version does not
+// exist.
+// - [machineerrors.MachineNotFound] when the machine specified by uuid  does
+// not exists.
 func (st *State) GetMachineTargetAgentVersion(ctx context.Context, uuid string) (coreagentbinary.Version, error) {
 	db, err := st.DB()
 	if err != nil {
 		return coreagentbinary.Version{}, errors.Capture(err)
 	}
 
-	info := machineAgentVersionInfo{MachineUUID: uuid}
+	info := machineTargetAgentVersionInfo{}
+	machineUUID := machineUUIDRef{
+		UUID: uuid,
+	}
 
 	stmt, err := st.Prepare(`
-SELECT av.target_version AS &machineAgentVersionInfo.target_version,
-       a.name AS &machineAgentVersionInfo.architecture_name
-FROM   agent_version AS av,
-       machine_agent_version AS mav
-JOIN   architecture AS a ON mav.architecture_id = a.id
-WHERE  mav.machine_uuid = $machineAgentVersionInfo.machine_uuid
-`, info)
+SELECT &machineTargetAgentVersionInfo.*
+FROM v_machine_target_agent_version
+WHERE machine_uuid = $machineUUIDRef.machine_uuid
+`, info, machineUUID)
 	if err != nil {
 		return coreagentbinary.Version{}, errors.Capture(err)
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		err := tx.Query(ctx, stmt, info).Get(&info)
+		err := st.checkMachineNotDead(ctx, tx, uuid)
+		if errors.Is(err, machineerrors.MachineNotFound) {
+			return errors.Errorf(
+				"machine %q does not exist", uuid,
+			).Add(machineerrors.MachineNotFound)
+		} else if err != nil && !errors.Is(err, machineerrors.MachineIsDead) {
+			return errors.Errorf(
+				"checking machine %q exists: %w", uuid, err,
+			)
+		}
```

Comment:
I think this deserves a comment. You are calling `checkMachineNotDead` then don't even care if it is dead.

You could, because of RI, assume the machine exists if the agent version exists for that machine.

Reply:
Done & Done